fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -1748,33 +1783,33 @@ void MultiVersionDatabase::DatabaseState::updateDatabase(Reference<IDatabase> ne
 		} else {
 			// For older clients that don't have an API to get the protocol version, we have to monitor it locally
 			try {
-				versionMonitorDb = MultiVersionApi::api->getLocalClient()->api->createDatabase(clusterFilePath.c_str());
+				versionMonitorDb = MultiVersionApi::api->getLocalClient()->api->createDatabase(connectionRecord);
 			} catch (Error& e) {
 				// We can't create a new database to monitor the cluster version. This means we will continue using the
 				// previous one, which should hopefully continue to work.
 				TraceEvent(SevWarnAlways, "FailedToCreateDatabaseForVersionMonitoring")
 				    .error(e)
-				    .detail("ClusterFilePath", clusterFilePath);
+				    .detail("ConnectionRecord", connectionRecord->toString());
 			}
 		}
 	} else {
 		// We don't have a database connection, so use the local client to monitor the protocol version
 		db = Reference<IDatabase>();
 		try {
-			versionMonitorDb = MultiVersionApi::api->getLocalClient()->api->createDatabase(clusterFilePath.c_str());
+			versionMonitorDb = MultiVersionApi::api->getLocalClient()->api->createDatabase(connectionRecord);
 		} catch (Error& e) {
 			// We can't create a new database to monitor the cluster version. This means we will continue using the
 			// previous one, which should hopefully continue to work.
 			TraceEvent(SevWarnAlways, "FailedToCreateDatabaseForVersionMonitoring")
 			    .error(e)
-			    .detail("ClusterFilePath", clusterFilePath);
+			    .detail("ConnectionRecord", connectionRecord->toString());
 		}
 	}
 	if (db.isValid() && dbProtocolVersion.present() && MultiVersionApi::apiVersionAtLeast(710)) {
-		auto updateResult =
-		    MultiVersionApi::api->updateClusterSharedStateMap(clusterFilePath, dbProtocolVersion.get(), db);
+		auto updateResult = MultiVersionApi::api->updateClusterSharedStateMap(
+		    connectionRecord->toString(), dbProtocolVersion.get(), db);
```

Comment:
I am wondering if it is still a reliable way to identify connections to the same cluster. The assumption was that an application can always use the same cluster file path for connections to the same cluster. If we have the memory-based connection string, how stable is it? What happens if coordinators change?

Reply:
I suppose neither version is really perfectly reliable (we can also open two files pointing to the same cluster), but the downfall of the memory version is that two separate records will never match. This is because the memory record string is based on a random ID created for each record and doesn't have anything to do with the connection string itself. It does preserve the existing behavior for file based connection records, where toString is basically just the filename of the cluster file.

Because the memory record version is exposed via internal APIs only, this may also be less of a concern for now. I'm not planning to use GRV caching for any connections using memory records.

A likely better fix for all of the various issues with memory and file based versions of this would be to pull a cluster ID from the cluster and use that to store the state. This probably isn't terribly difficult, we could sent it to each client as part of `ClientDBInfo` and if nothing else read it via the special key-space.

Reply:
I went ahead and implemented a version of this proposal where I fetch an ID from the cluster and use that to store the shared state.

Reply:
In the original design we considered the option fetching cluster ID and deliberately decided against it, because it adds a lot of overhead to the process of opening a database connection. Let's discuss that in a call.

Reply:
This same object is the one that includes the list of proxies, by the way, so I don't think there's much we can do with the cluster prior to having received it.

Reply:
If it receives the clusterID from the ClientDBInfo, then I think it should not affect performance significantly. I am however not sure about the overhead of retrieving it using a transaction on the special key space.